### PR TITLE
Enable batching in Ultralytics

### PR DESCRIPTION
**Pull request by manushreegangwar**
## What changes are proposed in this pull request?

Based on https://github.com/voxel51/fiftyone/pull/5834 to enable batching in ultralytics.

## How is this patch tested? If it is not, please explain why.

Apply model with batch sizes [1, 4, 8, 16]

```
import fiftyone as fo
import fiftyone.zoo as foz

dataset = fo.load_dataset("quickstart")

model_name = "yolo11n-coco-torch"

zoo_model = foz.load_zoo_model(
    model_name,
)

for batch in [1, 4, 8, 16]:
    label_field = f"{model_name}_collate_{batch}"
    dataset.apply_model(model, 
                        label_field=label_field,
                        batch_size=batch_size,
                        num_workers=8)


from ultralytics import YOLO

model = YOLO("yolo11n.pt")

for batch in [1, 4, 8, 16]:
    label_field = f"{model_name}_collate_{batch}_conv"
    dataset.apply_model(model, 
                        label_field=label_field,
                        batch_size=batch_size,
                        num_workers=8)
```

Checked correctness by comparing against direct YOLO inference where direct inference is done as below:

```
yolo_model = YOLO("yolo11n.pt")
# For each batch, compute direct inference results.
results = yolo_model(filepaths, batch=batch, verbose=False)

# Compare yolo inference results to  f"{model_name}_collate_{batch}" and  f"{model_name}_collate_{batch}_conv"

```

### Is this a user-facing change that should be mentioned in the release notes?

-   [ ] No. You can skip the rest of this section.
-   [ ] Yes. Give a description of this change to be included in the release
        notes for FiftyOne users.

(Details in 1-2 sentences. You can just refer to another PR with a description
if this PR is part of a larger change.)

### What areas of FiftyOne does this PR affect?

-   [ ] App: FiftyOne application changes
-   [ ] Build: Build and test infrastructure changes
-   [ ] Core: Core `fiftyone` Python library changes
-   [ ] Documentation: FiftyOne documentation changes
-   [x] Other

## Summary by CodeRabbit

- **Bug Fixes**
  - Improved image preprocessing for more consistent results and simplified batch handling in predictions.  
- **Refactor**
  - Updated internal image transformation and device transfer processes for better efficiency.  
  - Enhanced image augmentation with adaptive letterboxing based on input characteristics.  
- **New Features**
  - Added support for batch collation during prediction to improve batch processing flexibility.  
- **Documentation**
  - Clarified batch size configuration in Ultralytics model inference with examples for setting predictor batch size and rectangular resizing.

### CODE REVIEW

File: fiftyone/utils/ultralytics.py [hunk 1]
```diff
@@ -521,23 +529,33 @@ def _preprocess_img(self, img):
         if not isinstance(img, torch.Tensor):
             if isinstance(img, Image.Image):
                 img = img.convert("RGB")
-            orig_img = img
+            orig_img = np.array(img)
```
Comment (kaixi-wang) [inline fiftyone/utils/ultralytics.py:532]: are you intentionally wanting to use np.array instead of np.asarray? This returns a new array while the previous returns the input as a numpy array and so it more memory efficient/performant

File: fiftyone/utils/ultralytics.py [hunk 2]
```diff
@@ -557,30 +577,30 @@ def _predict_all(self, imgs):
...
 
         output = self._forward_pass(images)
 
         # This is required for Ultralytics post-processing.
         output["orig_imgs"] = orig_images
-        height, width = _get_image_dims(orig_images[0])
         output["imgs"] = images
 
+        width_height = []
+        for img in orig_images:
```
Comment (kaixi-wang) [inline fiftyone/utils/ultralytics.py:600]: the previous implementation only used the first image. now it seems like you are iterating over all images... for large numbers of images, couldn't this add a lot of time?

File: fiftyone/utils/ultralytics.py [hunk 3]
```diff
@@ -521,34 +530,49 @@ def _preprocess_img(self, img):
...
-        img = np.stack(self._model.predictor.pre_transform(img))
+        img = np.stack(self._pre_transform(img))
         img = img.transpose((0, 3, 1, 2))
         img = np.ascontiguousarray(img)
         img = torch.from_numpy(img)
         img = img.half() if self._model.predictor.model.fp16 else img.float()
         img /= 255
         return torch.squeeze(img, axis=0)
 
+    def _pre_transform(self, im):
```
Comment (jacobsela) [inline fiftyone/utils/ultralytics.py:550]: why does this have to be taken out here?

File: fiftyone/utils/ultralytics.py [hunk 4]
```diff
@@ -557,30 +581,30 @@ def _predict_all(self, imgs):
...
-            # Feed images as list
-            if self._output_processor is not None:
-                image = imgs[0]
-                height, width = _get_image_dims(image)
-        else:
-            # Feed images as stacked Tensor
-            if isinstance(images, (list, tuple)):
-                images = torch.stack(images)
+        # Feed images as stacked Tensor
+        if isinstance(images, (list, tuple)):
```
Comment (jacobsela) [inline fiftyone/utils/ultralytics.py:585]: this should be done in the collate

File: fiftyone/utils/ultralytics.py [hunk 5]
```diff
@@ -557,30 +581,30 @@ def _predict_all(self, imgs):
...
+            images = torch.stack(images)
 
-            height, width = images.size()[-2:]
+        self._model.predictor.setup_source(
```
Comment (jacobsela) [inline fiftyone/utils/ultralytics.py:588]: This kind of preprocessing shouldn't be in the predict.

**Review by jacobsela — CHANGES_REQUESTED** [review]
Haven't had a chance to run this yet. Will do so after changes are implemented.

Two general notes:
1. There are some preprocessing steps in `_predict_all` that should be moved out.
2. For testing, I want to make sure we are talking about the same thing when we say "direct yolo inference". Passing in a YOLO model to `apply_model` means it gets converted to the FO format and goes through the same inference code. A direct comparison would be comparing the results of `apply_model` to using the YOLO object directly, i.e. with the method `predict` of the base YOLO model. Please make sure to do the latter for testing.

File: fiftyone/utils/ultralytics.py (hunk 3 above)
Comment (manushreegangwar) [reply fiftyone/utils/ultralytics.py:553]: Good question. I had initially taken it out to set `auto` in LetterBox` to False but decided to make it dependent on `args.rect`. 

Seeing an odd behavior when I try to use `self._model.predictor.pre_transform` here. Added a todo comment.

File: fiftyone/utils/ultralytics.py (hunk 2 above)
Comment (manushreegangwar) [reply fiftyone/utils/ultralytics.py:603]: With batching enabled, each original image could be a different shape and we need shapes of each original image to scale the results correctly. Let me add this to collate.

File: fiftyone/utils/ultralytics.py [hunk 6]
```diff
@@ -624,7 +648,8 @@ class FiftyOneRTDETRModel(FiftyOneYOLOModel):
         config: a :class:`FiftyOneRTDETRModelConfig`
     """
 
-    pass
+    def _pre_transform(self, im):
```
[the remark this answers is not in the document]
Comment (manushreegangwar) [reply fiftyone/utils/ultralytics.py:650]: https://github.com/ultralytics/ultralytics/blob/main/ultralytics/models/rtdetr/predict.py#L72

I can remove this once https://github.com/voxel51/fiftyone/pull/5836/files#r2076046292 is resolved

File: fiftyone/utils/ultralytics.py (hunk 4 above)
Comment (manushreegangwar) [reply fiftyone/utils/ultralytics.py]: moved.

File: fiftyone/utils/ultralytics.py (hunk 5 above)
Comment (manushreegangwar) [reply fiftyone/utils/ultralytics.py]: I added a dummy batch initialization in place of this

File: fiftyone/utils/ultralytics.py [hunk 7]
```diff
@@ -521,34 +530,49 @@ def _preprocess_img(self, img):
...
+        config.output_processor_args["post_processor"] = None
+        return output_processor
 
     def _predict_all(self, imgs):
         if self._preprocess and self._transforms is not None:
             imgs = [self._transforms(img) for img in imgs]
+            if self.has_collate_fn:
+                imgs = self.collate_fn(imgs)
 
         if isinstance(imgs, list) and len(imgs) and isinstance(imgs[0], dict):
```
[the remark this answers is not in the document]
Comment (manushreegangwar) [reply fiftyone/utils/ultralytics.py]: Fixed.

File: fiftyone/utils/ultralytics.py (hunk 1 above)
Comment (manushreegangwar) [reply fiftyone/utils/ultralytics.py]: Good catch. Updated to `np.asarray` since the custom collate works without creating a copy.

**Review by jacobsela — CHANGES_REQUESTED** [review]
I'm getting incorrect labels for some samples with the following code. Is there some user error here?

```
dataset = fo.load_dataset("coco-2017").match_tags("validation")
model_name_fo = "yolo11l-coco-torch"
model = foz.load_zoo_model(model_name_fo)
batch_size = 32
dataset.apply_model(
    model,
    num_workers=64,
    batch_size=batch_size,
)
```

I'm comparing against
```
yolo_model_name = "yolo11l.pt"
yolo_model = ultralytics.YOLO(yolo_model_name)
for i in tqdm(range(0, len(dataset), batch_size)):
    batch = dataset[i : min(i + batch_size, len(dataset))]
    filepaths = batch.values("filepath")
    predictions = yolo_model.predict(filepaths)
    predictions = foultra.to_detections(predictions)
    batch.set_values("predictions_yolo", predictions)
```

For most of the samples the predictions are perfect, for around ~800 the predictions are completely off (like it looks like predictions for a different image). Am I doing something wrong?

**Comment by manushreegangwar** [comment]
> I'm getting incorrect labels for some samples with the following code. Is there some user error here?
> 
> ```
> dataset = fo.load_dataset("coco-2017").match_tags("validation")
> model_name_fo = "yolo11l-coco-torch"
> model = foz.load_zoo_model(model_name_fo)
> batch_size = 32
> dataset.apply_model(
>     model,
>     num_workers=64,
>     batch_size=batch_size,
> )
> ```
> 
> I'm comparing against
> 
> ```
> yolo_model_name = "yolo11l.pt"
> yolo_model = ultralytics.YOLO(yolo_model_name)
> for i in tqdm(range(0, len(dataset), batch_size)):
>     batch = dataset[i : min(i + batch_size, len(dataset))]
>     filepaths = batch.values("filepath")
>     predictions = yolo_model.predict(filepaths)
>     predictions = foultra.to_detections(predictions)
>     batch.set_values("predictions_yolo", predictions)
> ```
> 
> For most of the samples the predictions are perfect, for around ~800 the predictions are completely off (like it looks like predictions for a different image). Am I doing something wrong?

I have been running tests with batch size 1, 4, 8, 16. Let me try to reproduce this.

**Comment by manushreegangwar** [comment]
@jacobsela I was able to reproduce this issue on coco validation. 

There are 861/5000 frames where the predictions are off and it does look like there are from a different image.

I further tested that for batch size 1, FO model predictions match the yolo predictions `predictions_yolo`.

I don't have a solution for this yet but the inference is not working properly when batching on large datasets.

**Review by jacobsela — APPROVED** [review]
LGTM
